File: rust/src/bin/threaded-server.rs
use std::{io::{Read, Write}, net::{TcpListener, TcpStream}, os::windows::io::AsRawSocket, thread};

use utils;

enum ProcessingState {
    WaitForMsg,
    InMsg
}



fn main() {
    let args: Vec<String> = std::env::args().collect();    
    let mut port = 9090;

    if args.len() >= 2 {
        port = args[1].parse::<i32>().unwrap();
    }
    println!("Serving on port {}", port);
    
    let addr = format!("0.0.0.0:{}", port);
    let listener = TcpListener::bind(addr).unwrap();
    
    loop {
        let (connection, addr) = listener.accept().expect("ERROR on accept");
        utils::report_peer_connected(addr);

        thread::spawn(||{
            server_thread(connection);
        });        
    }
}

fn server_thread(stream: TcpStream) {
    let thread_id = thread::current().id();

    println!("Thrread {:?} created to handle connection with socket {}", thread_id, stream.as_raw_socket());
    serve_connection(stream);
    println!("Thread {:?} done", thread_id);
}

fn serve_connection(mut stream: std::net::TcpStream) -> () {
    if stream.write(b"*").expect("send") < 1 {
        panic!("send");
    }

    let mut state = ProcessingState::WaitForMsg;

    loop {
        let mut buf: [u8; 1024] = [0; 1024];
        let len = stream.read(&mut buf).expect("recv");
        if len == 0 {
            break;
        }

        for i in 0..len {
            let c = buf[i];
            match state {
                ProcessingState::WaitForMsg => {
                    if c == b'^' {
                        state = ProcessingState::InMsg;
                    }
                },
                ProcessingState::InMsg => {
                    if c == b'$' {
                        state = ProcessingState::WaitForMsg;
                    } else {
                        buf[i] += 1;
                        if stream.write(&buf[i..i+1]).expect("send error") < 1 {
                            return;
                        }
                    }
                }
            }
        }
    }

}

File: go/cmd/threaded-server/threaded-server.go
package main

import (
	"fmt"
	"net"
	"os"
	"runtime"
	"strconv"

	"github.com/zadenyip/Concurrency-Servers/utils"
)

type ProcessingState int

const (
	WaitForMsg ProcessingState = iota
	InMsg
)

func main() {
	args := os.Args
	port := 9090

	if len(args) >= 2 {
		var err error
		port, err = strconv.Atoi(args[1])
		if err != nil {
			panic(err)
		}
	}

	fmt.Printf("Serving on port %d\n", port)
	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
	if err != nil {
		panic(err)
	}

	for {
		connection, err := listener.Accept()
		if err != nil {
			panic("ERROR on accept\n")
		}

		utils.ReportPeerConnected(connection.RemoteAddr())

		go serverThread(connection)
	}

}

func serverThread(connection net.Conn) {
	id := runtime.NumGoroutine()
	rawConn, err := connection.(*net.TCPConn).SyscallConn()

	if err != nil {
		panic(err)
	}

	var fd uintptr
	rawConn.Control(func(fileDescriptor uintptr) {
		fd = fileDescriptor
	})

	fmt.Printf("Thread %d created to handle connection with socket %d\n", id, fd)
	serveConnection(connection)
	fmt.Printf("Thread %d done", id)
}

func serveConnection(connection net.Conn) {
	{
		len, err := connection.Write([]byte("*"))
		if len < 1 || err != nil {
			panic("send")
		}
	}

	state := WaitForMsg

	defer connection.Close()
	for {
		buf := [1024]byte{}

		len, err := connection.Read(buf[:])
		if err != nil && err.Error() != "EOF" {
			panic("recv")
		}

		if len == 0 {
			break
		}

		for i := 0; i < len; i++ {
			c := buf[i]
			switch state {
			case WaitForMsg:
				if c == '^' {
					state = InMsg
				}
			case InMsg:
				if c == '$' {
					state = WaitForMsg
				} else {
					buf[i] += 1
					len, err := connection.Write(buf[i : i+1])
					if len < 1 || err != nil {
						// stderr
						fmt.Fprintf(os.Stderr, "send error\n")
						connection.Close()
						return
					}
				}
			}
		}
	}

}
